File: src/distribution/dis.py3/FluxDriver.py
from scipy.optimize import linprog
import numpy as np
import sys

def FluxDriver(dataDictionary):
    # Get some stuff from the DF -
    STM = dataDictionary["stoichiometric_matrix"]
    OBJVECTOR = dataDictionary["objective_coefficient_array"]
    # Get Flux bounds sequence from the DF -
    FluxBounds = dataDictionary["default_flux_bounds_array"]
    # Get species bounds
    SpeciesBounds = dataDictionary["species_bounds_array"]
    # Get numbers
    NUM_Unbalanced = dataDictionary["extra_species_num"]
    NUM_Speices = len(SpeciesBounds)
    NUM_Balanced = NUM_Speices - NUM_Unbalanced

    # Equality constraints
    Aeq = STM[NUM_Unbalanced:NUM_Speices]
    # Formulate the bV -
    bVEq = np.zeros(NUM_Balanced)

    # Inequality constraints
    UNBALANCED_STM = STM[0:NUM_Unbalanced]
    bVLB = [-SpeciesBounds[i][0] for i in range(NUM_Unbalanced)]
    bVUB = [SpeciesBounds[i][1] for i in range(NUM_Unbalanced)]
    A = np.concatenate((UNBALANCED_STM, -1*UNBALANCED_STM))
    bV = np.concatenate((bVUB, bVLB))

    # Call the LP solver -
    opt = {"disp": True, "tol": 1.0E-8}
    res = linprog(OBJVECTOR, A_ub=A, b_ub=bV, A_eq=Aeq, b_eq=bVEq,
                  bounds=FluxBounds, options=opt)
    if res.status == 1:
        opt = {"disp": True, "tol": 1.0E-8, "maxiter": 10000}
        res = linprog(OBJVECTOR, A_ub=A, b_ub=bV, A_eq=Aeq, b_eq=bVEq,
                      bounds=FluxBounds, options=opt)
    if res.status == 2:
        sys.exit()

    UPTAKE = np.dot(UNBALANCED_STM, res.x)

    return res.fun, res.x, UPTAKE, res.status

File: src/distribution/dis.m/FluxDriver.m
% FluxDriver.m calling GLPK
function [FLOW,FVAL,UPTAKE,status] = FluxDriver(dataDictionary, MIN_MAX_FLAG)

	% Get some stuff from the DD -
	STM = dataDictionary('stoichiometric_matrix');
	[NUM_Species, NUM_Var] = size(STM);
	OBJVECTOR = dataDictionary('objective_coefficient_array');

	% Get Flux bounds from the DD -
	FluxBounds = dataDictionary('default_flux_bounds_array');
	FluxLB = FluxBounds(:,1);
	FluxUB = FluxBounds(:,2);
    % Formulate the VARTYPE vector (Flux)-
    VARTYPE = blanks(NUM_Var)';
    for id = 1:NUM_Var
        VARTYPE(id, 1) = 'C';
    end

	% Get species bounds
	SpeciesBounds = dataDictionary('species_bounds_array');
	% Get numbers
	NUM_Unbalanced = dataDictionary('extra_species_num');
	NUM_Balanced = NUM_Species - NUM_Unbalanced;

	% Equality constraints
	Aeq = STM((NUM_Unbalanced+1):NUM_Species, :);
	% Formulate the bV -
	bVEq = zeros(NUM_Balanced,1);

	% Inequality constraints
	UNBALANCED_STM = STM(1:NUM_Unbalanced, :);
	bVLB = SpeciesBounds(1:NUM_Unbalanced, 1);
	bVUB = SpeciesBounds(1:NUM_Unbalanced, 2);

    % Formulate the matrix of constraint coefficients
	A = [UNBALANCED_STM ; UNBALANCED_STM ; Aeq];
    % Formulate the column array containing the right-hand side value for
    % each constraint in the constraint matrix.
	bV = [bVUB ; bVLB; bVEq];
    % Formulate the CTYPE vector for each constraint
    CTYPE = blanks(length(bV))';
    for id = 1:NUM_Unbalanced
        CTYPE(id, 1) = 'U';
        CTYPE(id+NUM_Unbalanced, 1) = 'L';
    end
    offset = 2*NUM_Unbalanced;
    for id = (1+offset):(NUM_Balanced+offset)
        CTYPE(id) = 'S';
    end

	% Set the sense flag
	if (MIN_MAX_FLAG == -1)
		sense = -1;
	else
	 	sense = 1;
	end

	% Set some values for the optional parameters
	param.msglev = 3;
    param.tmlim = 3600; 

	% Call the glpk solver -
	[FLOW,FVAL,status, extra] = glpk(OBJVECTOR, A, bV, FluxLB, FluxUB, CTYPE, VARTYPE, sense, param);
    if (status == 109)
        fprintf('failed to solve in 1h, try 10h...');
        param.tmlim = 3600*10; 
        [FLOW,FVAL,status, extra] = glpk(OBJVECTOR, A, bV, FluxLB, FluxUB, CTYPE, VARTYPE, sense, param);
    end 
    if (status == 109) 
        fprintf('failed to find the optima after running for 10h');
    end 

	UPTAKE = UNBALANCED_STM*FLOW;
return;
